clang-tools-extra/clang-tidy/daedalean/DerivedClassesCheck.cpp

```diff
@@ -0,0 +1,218 @@
+//===--- DerivedClassesCheck.cpp - clang-tidy -----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "DerivedClassesCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+#include <sstream>
+#include <unordered_map>
+#include <set>
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+namespace {
+
+bool isInterface(const CXXRecordDecl *pDecl) {
+  if (!pDecl->field_empty()) {
+    return false;
+  }
+
+  if (pDecl->isEffectivelyFinal()) {
+    return false;
+  }
+
+  for (const auto & m : pDecl->methods()) {
+
+    if (llvm::isa<CXXDestructorDecl>(m) || m->isImplicit()) {
+      continue;
+    }
+    if (!m->isPure()) {
+      return false;
+    }
+  }
+
+  return !pDecl->methods().empty();
+}
+
+bool isFinal(const CXXRecordDecl *pDecl) {
+  const auto *Def = pDecl->getDefinition();
+  if (!Def)
+    return false;
+  return Def->hasAttr<FinalAttr>();
+
+}
+
+struct BaseRef {
+  std::string ref;
+  const CXXBaseSpecifier base;
+
+  std::string getName() const {
+    return base.getType()->getAsCXXRecordDecl()->getQualifiedNameAsString();
+  }
+
+  bool operator <(const BaseRef & other) const {
+    return ref < other.ref;
+  }
+};
+
+void walkBases(const CXXRecordDecl *pDecl, std::unordered_map<std::string, std::set<BaseRef>> & references) {
+  const auto ref = pDecl->getQualifiedNameAsString();
+  for (const auto base : pDecl->bases()) {
+    const auto decl = base.getType()->getAsCXXRecordDecl();
+    references[decl->getQualifiedNameAsString()].insert({ref, base});
+    walkBases(decl, references);
+  }
+}
+
+}
+
+void DerivedClassesCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(cxxRecordDecl(hasDefinition()).bind("x"), this);
+}
+
+void DerivedClassesCheck::check(const MatchFinder::MatchResult &Result) {
+
+  const auto *MatchedDecl = Result.Nodes.getNodeAs<CXXRecordDecl>("x");
+
+  if (!MatchedDecl || (MatchedDecl->methods().empty() && MatchedDecl->field_empty() && MatchedDecl->getNumBases() == 0)) {
+    // Skip empty class
+    return;
+  }
+
```

Comment:
It would help readability of this function a bit if you add a comment to every section which of the 7 sub-rules it checks.

```
1. Each interface MUST have a virtual default destructor.
2. Inheritance from base class with non-pure-virtual methods other than destructor and/or data members is forbidden.
3. Public inheritance MUST be used to implement interfaces.
4. Base class MUST be declared virtual if it is in diamond inheritance.
5. All entity names MUST be unique in all inherited interfaces. If two interfaces share the same methods they MUST extend the common base interface.
6. Virtual functions MUST contain exactly one of two specifiers: virtual for new function or final if function overrides method from base class.
7. If a class contains at least one non-pure-virtual method other than destructor it MUST be declared final.
```